File: civitas/src/civitas/SorpresaSalirCarcel.java
package civitas;


import java.util.ArrayList;

/**
 *
 * @author Usuario
 */
public class SorpresaSalirCarcel extends Sorpresa{

    private MazoSorpresa mazo;

    SorpresaSalirCarcel(MazoSorpresa mazo, String texto) {
        super(texto);
        this.mazo = mazo;
    }
    
    @Override
    public void aplicarAJugador(int actual, ArrayList<Jugador> todos) {
        if(jugadorCorrecto(actual, todos)) {
            informe(actual, todos);
            boolean tiene = false;
            for(int i = 0; i < todos.size(); i++) {
                if(todos.get(i).tieneSalvoconducto()) {
                    tiene = true;
                }
            }
            if(!tiene) {
                todos.get(actual).obtenerSalvoconducto(this);
                salirDelMazo();
            }
        }
    }
    
    /**
     * Sale del mazo si la sorpresa es salir de la carcel y la inhabilita.
     */
    void salirDelMazo() {
       mazo.inhabilitarCartaEspecial(this);
    }
    
    @Override
    public String toString() {
        String salida = "Mazo " + mazo + super.toString();
        return salida;
    }
    
    /**
     * Cuando la sorpresa SALIRCARCEL esta usada la habilita.
     */
    void usada() {
        mazo.habilitarCartaEspecial(this);
    }
    
}

File: civitasRuby/pkg/civitas-0.0.1/lib/sorpresa_salir_carcel.rb
# encoding:utf-8
module Civitas

  class SorpresaSalirCarcel < Sorpresa
    
    def initialize(texto, mazo)
      super(texto)
      @mazo = mazo
    end
    
    def aplicar_a_jugador(actual, todos)
      if jugador_correcto(actual, todos)
        super(actual, todos)
        tiene = false
        for i in (0..todos.size)
          if todos.at(i).tiene_salvoconducto
            tiene = true
          end
        end
        if !tiene
          todos.at(actual).obtener_salvoconducto(self)
          salir_del_mazo
        end
      end
    end
    
    def usada 
      @mazo.habilitar_carta_especial(self)
  end
  
  def salir_del_mazo
      @mazo.inhabilitar_carta_especial(self)
  end

   def to_string
      return super.to_string 
    end  
  
  public_class_method :new  
  end
end